python-package/xgboost/core.py

```diff
@@ -138,27 +138,42 @@ def c_array(ctype, values):
     return (ctype * len(values))(*values)
 
 
-def _maybe_from_pandas(data, feature_names, feature_types):
+def _maybe_from_pandas(data, label, feature_names, feature_types):
     """ Extract internal data from pd.DataFrame """
     try:
         import pandas as pd
     except ImportError:
-        return data, feature_names, feature_types
+        return data, label, feature_names, feature_types
 
     if not isinstance(data, pd.DataFrame):
-        return data, feature_names, feature_types
-
-    dtypes = data.dtypes
-    if not all(dtype.name in ('int64', 'float64', 'bool') for dtype in dtypes):
-        raise ValueError('DataFrame.dtypes must be int, float or bool')
+        return data, label, feature_names, feature_types
+
+    mapper = {'int8': 'int', 'int16': 'int', 'int32': 'int', 'int64': 'int',
+              'uint8': 'int', 'uint16': 'int', 'uint32': 'int', 'uint64': 'int',
+              'float16': 'float', 'float32': 'float', 'float64': 'float',
+              'bool': 'i'}
+
+    data_dtypes = data.dtypes
+    if not all(dtype.name in (mapper.keys()) for dtype in data_dtypes):
+        raise ValueError('DataFrame.dtypes for data must be int, float or bool')
+
+    if label is not None:
+        if isinstance(label, pd.DataFrame):
```

Comment:
This **works in current master**, because the ```DataFrame``` has no column name:
```python
X = pd.DataFrame({'COL1':[62, 31, 51, 19, 1.4]})
y = pd.DataFrame([1, 1, 0, 0, 1])
xgb.DMatrix(X, label=y)
```

Naming the ```DataFrame``` column **does not work in current master**, but will work after merge:
```python
X = pd.DataFrame({'COL1':[62, 31, 51, 19, 1.4]})
y = pd.DataFrame({'TARGET':[1, 1, 0, 0, 1]})
xgb.DMatrix(X, label=y)
```

If you would name the label column in your example, it still will not work after merge. 
```python
xgb.DMatrix(np.random.randn(100, 50), pd.DataFrame({'label':[1, 2, 3, 4, 5] * 20}, dtype=int))
```
The data is a ```numpy.ndarray``` and not a pandas ```DataFrame``` so it will skip the check. I assume people who use a ```DataFrame``` for the data can / will use a ```DataFrame``` for the label.